File: validate binary search tree.py
# TC-O(n)
# SC-O(1)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.validate(root,None,None)
    def validate(self,node, low, high):
        if node==None:
            return True
        if (low!=None and node.val<=low) or (high!=None and node.val>=high):
            return False
        return self.validate(node.left,low,node.val) and self.validate(node.right,node.val,high)

File: validate-binary-search-tree.js
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No, did it in class first.

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
let prev;
var isValidBST = function(root) {
    prev = null;
    return inOrderTraversal(root);
};

const inOrderTraversal = (root) => {
    // Base Case
    if (root === null) return true;
    
    // Logic
    if(inOrderTraversal(root.left) == false) return false;
    
    if (prev !== null && root.val <= prev.val) return false;
    
    prev = root;
    return inOrderTraversal(root.right);
}
